src/script/interpreter.cpp

```diff
@@ -1258,65 +1349,207 @@ class CTransactionSignatureSerializer
+                // provided as well, but if it wasn't, actual validation will fail anyway.
+                uses_bip341 = true;
+            } else {
+                // Treat every spend that's not known to native witness v1 as a Witness v0 spend. This branch may
+                // also be taken for unknown witness versions, but it is harmless, and being precise would require
+                // P2SH evaluation to find the redeemScript.
+                uses_bip143 = true;
+            }
+        }
     }
 
-    m_ready = true;
+    if (uses_bip143 || uses_bip341) {
+        // Computations shared between both sighash schemes.
+        m_bip341_prevouts_hash = GetPrevoutHash(txTo);
+        m_bip341_sequences_hash = GetSequenceHash(txTo);
+        m_bip341_outputs_hash = GetOutputsHash(txTo);
+    }
+    if (uses_bip143) {
+        hashPrevouts = SHA256Uint256(m_bip341_prevouts_hash);
+        hashSequence = SHA256Uint256(m_bip341_sequences_hash);
+        hashOutputs = SHA256Uint256(m_bip341_outputs_hash);
+        m_bip143_ready = true;
+    }
+    if (uses_bip341) {
+        m_bip341_spent_amounts_hash = GetSpentAmountsHash(m_spent_outputs);
+        m_bip341_spent_scripts_hash = GetSpentScriptsHash(m_spent_outputs);
+        m_bip341_ready = true;
+    }
 }
 
 template <class T>
 PrecomputedTransactionData::PrecomputedTransactionData(const T& txTo)
 {
-    Init(txTo);
+    Init(txTo, {});
 }
 
 // explicit instantiation
-template void PrecomputedTransactionData::Init(const CTransaction& txTo);
-template void PrecomputedTransactionData::Init(const CMutableTransaction& txTo);
+template void PrecomputedTransactionData::Init(const CTransaction& txTo, std::vector<CTxOut> spent_outputs);
+template void PrecomputedTransactionData::Init(const CMutableTransaction& txTo, std::vector<CTxOut> spent_outputs);
 template PrecomputedTransactionData::PrecomputedTransactionData(const CTransaction& txTo);
 template PrecomputedTransactionData::PrecomputedTransactionData(const CMutableTransaction& txTo);
 
+static const CHashWriter HASHER_TAPSIGHASH = TaggedHash("TapSighash");
+static const CHashWriter HASHER_TAPLEAF = TaggedHash("TapLeaf");
+static const CHashWriter HASHER_TAPBRANCH = TaggedHash("TapBranch");
+static const CHashWriter HASHER_TAPTWEAK = TaggedHash("TapTweak");
+
+template<typename T>
+bool SignatureHashSchnorr(uint256& hash_out, const ScriptExecutionData& execdata, const T& tx_to, const uint32_t in_pos, const uint8_t hash_type, const SigVersion sigversion, const PrecomputedTransactionData* cache)
+{
+    uint8_t ext_flag;
+    switch (sigversion) {
+    case SigVersion::TAPROOT:
+        ext_flag = 0;
+        break;
+    case SigVersion::TAPSCRIPT:
+        ext_flag = 1;
+        break;
+    default:
+        assert(false);
+    }
+    assert(in_pos < tx_to.vin.size());
+    assert(cache != nullptr && cache->m_bip341_ready && cache->m_spent_outputs_ready);
```

Comment:
If I want to taproot-sign a transaction, it will not have any ScriptWitness data when I try to obtain the signature hash, which means `m_bip341_ready` will not be set after the `PrecomputedTransactionData::Init()` call has been made.

This means I have to put in junk data in the vin's scriptwitness in order to obtain the sighash, which seems off.

Reply:
Will you be using a cache when signing a tx? I figured no.

Reply:
I need the sighash to sign the transaction, and I need the cache to get the sighash.

Reply:
There is no signing code yet, so this may go in a number of different ways, but I was imagining that the cache would indeed be used for signing as well, as it becomes increasingly inconvenient to pass all arguments (including all spent UTXOs now!) along down into the signing function.

In fact, I think after the change to use the cache to pass that information along, we can stop passing scriptPubKey/amount/txFrom to `SignSignature`... but that's for much later.

@kallewoof It's indeed ugly in the signing case that the scriptWitness is used to determine if a transaction is a witness input. I'll change that for now to just default to BIP143 caching unless it's clearly a BIP341 spend, instead of relying on there being a non-empty witness.

Reply:
@sipa I worked around it on my end, so feel free to leave it as is until a decision is made on how signing should work!